### PR TITLE
8286957: Held monitor count

### PR DESCRIPTION
**Pull request by robehn**
The current implementation do not count all monitor enter, counts high up in abstraction and causes a performance regression on aarch64 with some benchmarks due to C2 changes.

This change makes the counting exact by pushing the counting down in the abstraction.
The additional JNI counter is strictly not needed, but enables us to figure out if we have monitors "on stack".

An uncontended lock plus unlock is 1 ns (21.5 -> 22.5) slower in C2 compiled code on x64 with the additional increment and decrement.

Fixed aarch64, x64, x86 and zero.

Passes t1-8

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286957](https://bugs.openjdk.org/browse/JDK-8286957): Held monitor count

### Reviewers
 * [Ron Pressler](https://openjdk.org/census#rpressler) (@pron - Committer) ⚠️ Review applies to [fb49d0e5](https://git.openjdk.org/jdk/pull/8945/files/fb49d0e5c31225a6446597125e900c5e172acc73)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/8945/head:pull/8945` \
`$ git checkout pull/8945`

Update a local copy of the PR: \
`$ git checkout pull/8945` \
`$ git pull https://git.openjdk.org/jdk pull/8945/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8945`

View PR using the GUI difftool: \
`$ git pr show -t 8945`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/8945.diff">https://git.openjdk.org/jdk/pull/8945.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/interpreter/zero/bytecodeInterpreter.cpp [hunk 1]
```diff
@@ -626,14 +626,19 @@ void BytecodeInterpreter::run(interpreterState istate) {
         markWord displaced = rcvr->mark().set_unlocked();
         mon->lock()->set_displaced_header(displaced);
         bool call_vm = UseHeavyMonitors;
+        bool inc_mon_count = true;
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/share/interpreter/zero/bytecodeInterpreter.cpp:629]: Please do not use `mont` as an abbreviation for monitor - use `mon` or `monitor`. Thanks.

[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/share/interpreter/zero/bytecodeInterpreter.cpp]: Thanks, fixed.

**Comment by robehn** [comment]
Thank you @pron !

**Review by fisk — CHANGES_REQUESTED** [review]
Looks good. We might however want the counter to be 64 bit so we don't have to think about overflows. I suppose nasty JNI code could lock the entire heap and then unlock it.

**Comment by robehn** [comment]
Starting retesting.

**Comment by robehn** [comment]
Incremental changes passed t1-7.

@pron and @fisk can you please have a look if the incremental changes are okay?

And additionally can @pron please take careful look at continuationEntry.hpp and continuationFreezeThaw.cpp

Thanks

**Comment by robehn** [comment]
@fisk thank you!

@pron thank you!

It passed loom t1-5 also.

**Pull request closed without merging**